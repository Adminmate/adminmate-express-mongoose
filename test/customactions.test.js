// Hide console.log
global.console = {
  log: jest.fn(),
  // Keep native behaviour for other methods
  error: console.error,
  warn: console.warn,
  info: console.info,
  debug: console.debug,
};

const supertest = require('supertest');
const jwt = require('jwt-simple');

// Include the app
const app = require('./app.js');

// Endpoint prefix
const prefix = '/adminmate/api';

// Generate the admin token
const adminToken = jwt.encode({
  exp_date: Date.now() + 1000
}, 'authkey_secret');

// Before all
beforeAll(done => {
  done();
});

// Custom actions
describe('Testing GET /api/models/customactions', () => {
  it('should return a 403 http response', async () => {
    // Make request
    const response = await supertest(app)
      .get(prefix + '/models/customactions');

    // Check response
    expect(response.status).toBe(403);
    expect(response.body.code).toBe('not_authorized');
  });

  it('should return a 200 http response', async () => {
    // Make request
    const response = await supertest(app)
      .get(prefix + '/models/customactions')
      .set('x-access-token', adminToken);

    // Check response
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});

// Custom actions
describe('Testing GET /api/models/:model/customactions', () => {
  it('should return a 403 http response', async () => {
    // Make request
    const response = await supertest(app)
      .get(prefix + '/models/users/customactions');

    // Check response
    expect(response.status).toBe(403);
    expect(response.body.code).toBe('not_authorized');
  });

  it('should return a 403 http response', async () => {
    // Make request
    const response = await supertest(app)
      .get(prefix + '/models/users/customactions')
      .set('x-access-token', adminToken)
      .query({ ids: [], target: '' })

    // Check response
    expect(response.status).toBe(403);
    expect(response.body.message).toBe('Invalid request');
  });

  it('should return a 200 http response', async () => {
    // Make request
    const response = await supertest(app)
      .get(prefix + '/models/users/customactions')
      .set('x-access-token', adminToken)
      .query({ ids: [9], target: 'item' })

    // Check response
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});

// After all
afterAll(done => {
  done();
})