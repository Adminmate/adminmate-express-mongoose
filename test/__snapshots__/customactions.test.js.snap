// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GET /api/models/:model/customactions should return a 200 http response 1`] = `
Object {
  "list": Array [
    Object {
      "code": "delete",
      "label": "Delete item",
    },
  ],
}
`;

exports[`Testing GET /api/models/customactions should return a 200 http response 1`] = `
Object {
  "list": Array [],
}
`;
